book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,228 @@
+allows the validator to prune its state and also calculate leader schedules as
+it moves across epochs.
+
+4) Wait to observe a canary transaction in a bank that some threshold of
+trusted validator stake has voted on. Call this trusted bank `T`.
+
+5) Figure out what banks the validator is not locked out from based on it's
+locktower. Every validator persists its locktower state and must consult this
+state in order to boot safely and resume from a snapshot without being slashed.
+From this locktower state and the ancestry information embedded in the
+snapshot, a validiator can derive which banks, are "safe" (See the `Determining
+Vote Safety From a Snapshot and Locktower` section for more detals) to vote
+for.
+
+6) Wait until it is "safe" to vote on the snapshot root `S`.
+
+7) Start voting for any descendant of a trusted bank `T` that is also "safe"
+
+
+For the sections below:
+
+Assume an arbitrary locktower state `L`, a snapshot root `S`, and a trusted
+bank slot `T` (defined in step 4 of the `Snapshot Verification Overview`
+section).
+
+Define `L_i` to be: The slot of the ith vote in `L`.
+
+Define the function `is_ancestor(a, b)` to be: Given a slot `a` and a slot `b`
+will return true if `a` is an ancestor of `b`.
+
+Define the function `is_locked(a, b)` to be: `a + a.lockout <= b`
+
+## Determining Vote Safety From a Snapshot and Locktower
+
+Define the "safety" condition to be: 
+
+Given a `BankForks`, a validator can run some procedure to determine whether it
+can vote on a frozen bank `T` (frozen banks must be present in
+`BankForks.banks`), without violating the lockouts of any vote in `L`. The
+procedure for this is:
+
+1) If `T` < last locktower vote, return false,
+
+2) Apply `T` to locktower state `L`. Pop off all votes `S_i` where
+`!is_locked(S_i, T)`. `T` must now be the top of the tower.
+
+3) If for any remaining vote `L_i` in `L` `!is_ancestor(L_i, T)`, return false.
+
+4) Otherwise, return true.
+
+## Achieving Safety
+
+Define the "Safety Criteria" to be: Given a `BankForks` any descendant of `S`,
+`S_d` that has a bank `B_d` that is present in `BankForks.banks`, and any slot
+`L_i` in locktower, the validator is able to determine `is_ancestor(L_i, S_d)`.
+
+Assume the "Safety Criteria" is true, we show we can then achieve the "safety"
+condition:
+
+**Proof:** A validator wants to determine whether it can vote on `T`. `T` must
+be a descendant of `S` because the validator does not play any state for
+non-descendants of `S` when booting from the snapshot. From assuming "Safety
+Criteria" above, this means for any `L_i` we can determine `is_ancestor(L_i,
+T)`. This means we have all the tools to run the algorithm from the safety
+definition.
+
+Thus to achieve safety, we want to design the snapshotting system such that the
+"Safety Criteria" is met.
+
+## Implementing "Safety Criteria"
+
+"Range Assumption": In our implementation, we assume that at the time the
```

Comment:
We can get rid of this range assumption now that you have an explicit case for it